File: scala-functional/cipher.scala
// Caesar Cipher - Scala (functional)
// Author: John Paul Welsh
object Cipher {
    
  def encrypt(input: List[Char], shiftAmt: Int): List[Char] = {
            
    def shiftWithBounds(num: Int, amt: Int): Int = {
    val temp = num + amt
    if (num == 32)      num
    else if (temp < 65) 91 - (65 % temp)
    else if (temp > 90) 64 + (temp % 90)
    else                temp
    }

    val inputInts = input.map(c => c.asInstanceOf[Int])
    val shiftedInts = inputInts.map(n => shiftWithBounds(n, shiftAmt))
    val backToChar = shiftedInts.map(n => n.asInstanceOf[Char])

    backToChar
  }

  def decrypt(input: List[Char], shiftAmt: Int): List[Char] =
    encrypt(input, -shiftAmt)

  def solve(input: List[Char], amt: Int): Unit = {
    if (amt >= 0) {
      println("Solve " + amt + ": " + decrypt(input, amt % 26).mkString)
      solve(input, amt - 1)
    }
  }

  def main(args: Array[String]) {
    val input = "John Paul Welsh".toUpperCase
    val shiftAmt = 15
    val maxShiftVal = 26

    println("The input string is " + input)
    println("The shift amount for encrypt and decrypt is " + shiftAmt)
    println("The maximum shift value for solve is " + maxShiftVal)
    println("Encrypted: " + encrypt(input.toList, shiftAmt % 26).mkString)
    println("Decrypted: " + decrypt(input.toList, shiftAmt % 26).mkString)
    solve(input.toList, maxShiftVal)
  }
}

File: erlang/cipher.erl
% Caesar Cipher - Erlang
% Author: John Paul Welsh
-module(cipher).
-export([shift/2]).
-export([encrypt/2]).
-export([decrypt/2]).
-export([solve/2]).
-export([display/0]).

% Shifts a number, with bounds-checking, and ignoring spaces.
shift(Num, Amt) ->
	Temp = Num+Amt,
	if
		Num =:= 32 -> Num;
		Temp < 65 -> (91 - (65 rem Temp));
		Temp > 90 -> (64 + (Temp rem 90));
		true -> Temp
	end.

% Shifts the head of the current input, puts that in a list, and
% concatenates that list with another one containing a recursive
% call to the function.
encrypt(Input, Amt) ->
	if
		length(Input) < 1 -> [];
		true -> ([shift(hd(Input), Amt)] ++ [encrypt(tl(Input), Amt)])
	end.

% Encrypts with a negative number.
decrypt(Input, Amt) ->
	encrypt(Input, -Amt).

% Prints the decryption of the current amount, then recursively
% calls the function with one smaller shift amount, down to 0.
solve(Input, Max) ->
	if
		Max < 0 -> [];
		true -> io:format("Solve ~w: ~s~n", [Max, decrypt(Input, Max)]),
			solve(Input, Max-1)
	end.

% Prints everything to the screen: basically a 'main' function.
display() ->
	InputString = "JOHN PAUL WELSH",
	ShiftAmt = 18,
	MaxShift = 26,

	io:format("The input string is ~s~n", [InputString]),
	io:format("The shift amount for encrypt and decrypt is ~w~n", [ShiftAmt]),
	io:format("The maximum shift amount for solve is ~w~n", [MaxShift]),
	io:format("Encrypted: ~s~n", [encrypt(InputString, ShiftAmt)]),
	io:format("Decrypted: ~s~n", [decrypt(InputString, ShiftAmt)]),
	solve(InputString, MaxShift).